core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala

```diff
@@ -0,0 +1,214 @@
+                                                + "znodes as part of the process of setting up ZooKeeper "
+                                                + "authentication.")
+  }
+  
+  def run(args: Array[String]) {
+    var jaasFile = System.getProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM)
+    val parser = new OptionParser()
+
+    val goOpt = parser.accepts("kafka.go", "Indicates whether to make this cluster secure or unsecure. The "
+        + " options are 'secure' and 'unsecure'").withRequiredArg().ofType(classOf[String])
+    val jaasFileOpt = parser.accepts("jaas.file", "JAAS Config file.").withOptionalArg().ofType(classOf[String])
+    val zkUrlOpt = parser.accepts("zookeeper.connect", "Sets the ZooKeeper connect string (ensemble). This parameter " +
+      "takes a comma-separated list of host:port pairs.").withRequiredArg().defaultsTo("localhost:2181").
+      ofType(classOf[String])
+    val zkSessionTimeoutOpt = parser.accepts("zookeeper.session.timeout", "Sets the ZooKeeper session timeout.").
+      withRequiredArg().defaultsTo("30000").ofType(classOf[java.lang.Integer])
+    val zkConnectionTimeoutOpt = parser.accepts("zookeeper.connection.timeout", "Sets the ZooKeeper connection timeout.").
+      withRequiredArg().defaultsTo("30000").ofType(classOf[java.lang.Integer])
+    val helpOpt = parser.accepts("help", "Print usage information.")
+
+    val options = parser.parse(args : _*)
+    if(options.has(helpOpt))
+      outputUsage(parser)
+
+    if ((jaasFile == null) && !options.has(jaasFileOpt)) {
+      error("No JAAS configuration file has been specified. Please make sure that you have set either "
+            + "the system property %s or the option %s".format(JaasUtils.JAVA_LOGIN_CONFIG_PARAM, "--jaas.file"))
+      throw new IllegalArgumentException("Incorrect configuration")
+    }
+    
+    if (jaasFile == null) {
+      jaasFile = options.valueOf(jaasFileOpt)
+      System.setProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM, jaasFile)
+    }
+    
+    if (!JaasUtils.isZkSecurityEnabled(jaasFile)) {
+      error("Security isn't enabled, most likely the file isn't set properly: %s".format(jaasFile))
+      throw new IllegalArgumentException("Incorrect configuration") 
+    }
+
+    var goSecure: Boolean = false
+    val go = options.valueOf(goOpt) match {
+      case "secure" =>
+        goSecure = true
+      case "unsecure" =>
+        goSecure = false
+      case _ =>
+        outputUsage(parser)
+    }
+    val zkUrl = options.valueOf(zkUrlOpt)
+    val zkSessionTimeout = options.valueOf(zkSessionTimeoutOpt).intValue
+    val zkConnectionTimeout = options.valueOf(zkConnectionTimeoutOpt).intValue
+    val zkUtils = ZkUtils.apply(zkUrl, zkSessionTimeout, zkConnectionTimeout, goSecure)
+
+    for (path <- zkUtils.securePersistentZkPaths) {
+      zkUtils.makeSurePersistentPathExists(path)
+      setAclsRecursively(zkUtils, path)
+    }
+
+    while(counter.get > 0 && exception != null) {
```

Comment:
Instead of this, you could return the Futures from `setAclsRecursively`, call `Future.sequence` and then use `Await.result`.